File: gearshift2/src/main/java/org/sugr/gearshift/viewmodel/RetainedFragment.java
package org.sugr.gearshift.viewmodel;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.SharedPreferences;
import android.os.Bundle;

import org.sugr.gearshift.App;

public class RetainedFragment<VM extends RetainedViewModel<T>, T> extends Fragment {
    private VM viewModel;

    // Interface for creating a view model with a lambda
    public interface Factory<VM extends RetainedViewModel<T>, T> {
        VM create(SharedPreferences prefs);
    }

    @Override public void onCreate(Bundle state) {
        super.onCreate(state);
        setRetainInstance(true);
    }

    @Override public void onDestroy() {
        super.onDestroy();

        if (viewModel != null) {
            viewModel.onDestroy();
        }
    }

    public void setViewModel(VM viewModel) {
        this.viewModel = viewModel;
    }

    public VM getViewModel() {
        return viewModel;
    }

    @SuppressWarnings("unchecked")
    public static <VM extends RetainedViewModel<T>, T> VM getViewModel(FragmentManager fm,
                                                                       String tag,
                                                                       Factory<VM, T> factory) {

        RetainedFragment<VM, T> fragment =
                (RetainedFragment<VM, T>) fm.findFragmentByTag(tag);

        if (fragment == null) {
            fragment = new RetainedFragment<>();
            fragment.setViewModel(factory.create(App.defaultPreferences()));

            // TODO: commit -> commitNow (support v24)
            fm.beginTransaction().add(fragment, tag).commit();
        }

        return fragment.getViewModel();
    }
}

File: gearshift2/src/main/kotlin/org/sugr/gearshift/viewmodel/RetainedFragment.kt
package org.sugr.gearshift.viewmodel

import android.content.SharedPreferences
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import org.sugr.gearshift.defaultPreferences

class RetainedFragment<VM : RetainedViewModel<*>> : Fragment() {
    var viewModel: VM? = null

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        retainInstance = true
    }

    override fun onDestroy() {
        super.onDestroy()

        if (viewModel != null) {
            viewModel!!.onDestroy()
        }
    }
}

inline fun <reified VM : RetainedViewModel<*>> viewModelFrom(fm: FragmentManager,
                                                             tag: String = VM::class.java.toString(),
                                                             factory: (tag: String, prefs: SharedPreferences) -> VM): VM {
    var fragment = fm.findFragmentByTag(tag) as? RetainedFragment<VM>

    if (fragment == null) {
        fragment = RetainedFragment<VM>()
        fragment.viewModel = factory(tag, defaultPreferences())

        fm.beginTransaction().add(fragment, tag).commitNow()
    }

    return fragment.viewModel as VM
}

fun <VM : RetainedViewModel<*>> destroyViewModel(fm: FragmentManager,
                                                    viewModel: VM) {

    val fragment = fm.findFragmentByTag(viewModel.tag) as? RetainedFragment<*>

    if (fragment != null) {
        fm.beginTransaction().remove(fragment).commitNow()
    }
}
